### DISCUSSION TITLE
How do I give a sequence symbol precedence over a regex symbol

### DISCUSSION THREAD

**Discussion by Hubro** [tree-sitter/tree-sitter#1287] [Q&A] [answered]
I have a grammar that is working very nicely:

```javascript
argument: $ => choice(
    $.built_in_type,
    $.node_identifier,
    $.integer,
    $.boolean,
    $.string,
    $.string_concatenation,
    $.date,
    $.range,
    $.keypath,
    $.yang_version,
    $.glob,
    $.unquoted_string,
),

// ...

integer: $ => choice(
    /-?\d+/,
    /-?0[xX][a-zA-Z0-9]+/
),

range: $ => choice(
    $.unquoted_range,
    $.quoted_range,
),

unquoted_range: $ => /\d+\.\.\d+/,
quoted_range: $ => /"\d+\.\.\d+"/,

// Unquoted strings are not explained in the ABNF grammar, but going by
// all the examples in yang-modules, it seems like it can contain any
// symbol except for whitespace, semicolons, quotes and curly brackets.
unquoted_string: $ => /[^\s;"'{}]+/,

// ...
```

However, I want to improve the parsing of the `$.range` symbol to parse the starting number, the dots and the end number separately:

```scm
(range (unquoted_range (start) (dots) (end)))
```

The problem is, the moment I replace the single regex for a `seq`:

```javascript
unquoted_range: $ => seq(
    alias(/\d+/, $.start),
    alias('..', $.dots),
    alias(/\d+/, $.end),
),
```

Tree-sitter will now prefer to parse the symbol as a `$.unquoted_string` instead. This makes sense, since the `$.unquoted_string` symbol is simpler (a single regex) and matches the entire token.

Just to test, I commented out the `$.unquoted_string` symbol, and Tree-sitter now chose to parse the text as the `$.integer` symbol, and then throw an error because it didn't recognize the dots.

If I additionally comment out `$.integer`, now it parses the text as an `$.unquoted_range`.

It seems to me like Tree-sitter will *never* choose to parse a chunk of text as a *sequence* symbol if there are any *regex* or *string* symbols that also match the text.

Is there any way to make Tree-sitter prioritize a sequence over a single regex?

---

If I wrap the sequence in `token` it works, but then I lose the nested symbols (start, dots and end)

**Comment by maxbrunsfeld** [comment]
I think you can give `unquoted_string` a lower precedence at the lexical level:

```js
unquoted_string: token(prec(-1, /[^\s;"'{}]+/)),
```

For more information, see the docs on conflicting tokens: https://tree-sitter.github.io/tree-sitter/creating-parsers#conflicting-tokens.

**Reply by Hubro** [reply]
I thought about doing something like that as well, but this causes Tree-sitter to parse the text as anything other than an unquoted string, even when that leads to errors. For example:

```yang
default 1.2.3.4;
```

The argument "1.2.3.4" will be parsed as an `$.integer`:

```javascript
integer: $ => choice(
    /-?\d+/,
    /-?0[xX][a-zA-Z0-9]+/
),
```

And the syntax tree will end in an error because Tree-sitter doesn't know what to do with the remaining ".2.3.4".

I feel like `prec.right` should be used here somehow, since I want the longest match :thinking: I haven't figured out how to apply it yet though.

**Reply by ahlinc** [reply]
As I saw in `tree-sitter parse -d` command with the debug log when the `token(prec(-1, ...))` applied to a regexp, it starts to behave like it's non greedy.

**Comment by ahlinc** [comment, marked as answer]
@Hubro First of all see the **Edit** section for my comment here: https://github.com/tree-sitter/tree-sitter/discussions/1269#discussioncomment-1010753. It's tricky to always remember all tiny details and I accidentally misled you, sorry.
You need to rollback your `_sep` token to the `/\s+/` definition, it's important.
Than try to add lexer precedence to your `start` rule and resolve contention between your terminal regular expressions. To be honest I wasn't able to do it quickly in my experiment, but I'm unaware with your grammar. May be without an external scanner it's unsolvable.

**Reply by Hubro** [reply]
@ahlinc No worries, thanks for the update!

It doesn't seem like lexical precedence is going to be sufficient. If I apply lexical precedence to the "start" rule, then all numbers in the document are going to be parsed as the "start" symbol, even if it makes absolutely no sense in the context, and even if the "start" symbol isn't even valid in that position in the syntax tree. It will parse "1.1.1.1" as "(start) (start) (start) (start)", which is absolute nonsense.

If I apply lexical precedence to the "integer" rule, then all numbers are parsed as integers instead, even if the resulting syntax tree is erroneous.

I think I have to instruct Tree-sitter to *try a path* (parse the argument as an integer) then proceed, and if it finds unexpected text afterwards, like `..`, it has to go *back* to the argument parsing and try something else, like a "range". It's not enough to just look at the number itself to know if its a "start" or "integer" symbol, it's the following symbols that decide that.

Is there any way to do this? So far I have been completely unsuccessful.

**Reply by maxbrunsfeld** [reply]
Regardless of lexical precedence, the parser will never try to match a token that isn’t valid in the current context, according to your grammar.

unless a parse error has already occurred.

**Reply by Hubro** [reply]
Perhaps "in the context" was a poor choice of words. It parses `(start) (start) (start) (start)`, which is *never* valid according to my grammar. There can never be two "start" symbols in a row. A "start" symbol can only ever be followed by a "dots" symbol, then an "end" symbol. There are other valid paths for the parser to take when parsing the argument, but it never tries. It picks the immediate next symbol with the highest precedence ("start") and proceeds, never looking ahead or going back.

Is there any mechanism in Tree-sitter for parsing the next sequence of symbols based on which choice results in a correct syntax tree?

**Reply by Hubro** [reply]
@maxbrunsfeld @ahlinc 

Here is a simplified case to summarize the problem:

I want to parse this statement: `<keyword> <argument>;`

Here are a few valid statements:

```yang
default 10;   // (statement (keyword) (argument (integer))
default 10..100;   // (statement (keyword) (argument (range))
default 127.0.0.1;   // (statement (keyword) (argument (unquoted_string))
```

As long as `$.argument` is a single token, this is easy! I just use a `choice` with each child symbol defined in a logical order of precedence (integer > range > unquoted_string).

However, if I want to parse `$.range` as `(argument (range (start) (dots) (end)))` for better syntax highlighting, everything breaks.

- Because `$.unquoted_string` gives a *longer* match than `$.start` ("10..100" vs "10") it will take precedence over `$.start`:

  ```yang
  default 10..100;   // (statement (keyword) (argument (unquoted_string))
  ```

- If I give `$.start` lexical precedence using `token(prec(1, ...))` then it takes precedence over everything, regardless of whether it completely breaks the syntax tree:

  ```yang
  default 10;   // (ERROR statement (keyword) (argument (start))
  default 10..100;   // (statement (keyword) (argument (range (start) (dots) (end)))
  default 127.0.0.1;   // (ERROR statement (keyword) (argument (start) (start) (start) (start))
  ```

So the problem is, if Tree-sitter has the option between lexing the argument as a single token or a sequence of tokens, it will always match it as a single token. If I give the *first* token of the sequence a higher lexical precedence, then Tree-sitter will *always* pick that token even if the following symbols in the sequence don't match anything, resulting in an erroneous syntax tree.

I can imagine two theoretical solutions, but I have no clue if Tree-sitter is capable of either of them:
- Instruct Tree-sitter to try parsing up until (for example) the end of the `word`. If it encounters an error, go back and try the next alternative path, in order of precedence.
- Instruct Tree-sitter to lex a symbol as a single token, but parse it as multiple tokens. For example:

  ```javascript
  range: $ => lex_as_token(seq(
      alias(/\d+/), $.start),
      alias('..', $.dots),
      alias(/\d+/, $.end),
  )),
  ```

  That way `$.range` could compete with symbols like `$.unquoted_string` lexically, but still be parsed as multiple tokens in the syntax tree.

Is there currently any way to achieve what I'm trying to do?

**Reply by maxbrunsfeld** [reply]
Ok, there shouldn't be a separate *token* called `start`. The start of a range expression is not something that the lexer can distinguish by itself (because it depends what token comes *next* in the sequence).

You need to have a more generic token called `integer` or `number` or something, and it should be used in both of these places:

```c
default 10;      // (statement (range (number) (number)))
default 10..100; // (statement (number))
```

In general, I would recommend looking at some other Tree-sitter grammars for popular languages. What you're trying to do is very standard and there are similar constructs in most of the existing grammars.

**EDIT**

Looking at your original grammar, it seems like you were close. You already have a token called `integer`. Just use that in `range`. Does that make sense? If you try to describe them as two separate tokens, it's not going to be possible for the lexer to know which one to produce.

**Reply by Hubro** [reply]
@maxbrunsfeld Ah of course, that makes sense :man_facepalming: That was dumb. But I don't think it's related to the problem. If I define `$.range` like:

```javascript
range: $ => seq(
  $.integer,
  '..',
  $.integer,
),
```

The lexer will still prefer to interpret the range as `$.unquoted_string`, because that is a longer match than the first `$.integer` in the `$.range` sequence.

```c
default 10;   // (statement (keyword) (argument (integer))
default 10..100;   // (statement (keyword) (argument (unquoted_string))
default 127.0.0.1;   // (statement (keyword) (argument (unquoted_string))
```

**Reply by maxbrunsfeld** [reply]
Do you have the `token(prec(-1, ...))` wrapping the `unquoted_string` rule?

**Reply by Hubro** [reply]
@maxbrunsfeld Unfortunately, if I do that, any argument that starts with a number will be parsed as a number, even if it's part of a text string, for example:

```c
default 10;   // (statement (keyword) (argument (integer))
default 10..100;   // (statement (keyword) (argument (range (integer) (integer))))
default 127.0.0.1;   // (ERROR statement (keyword) (argument (integer) (string_fragment))
```

```javascript
unquoted_string: $ => token(prec(-1, /[^\s;"'{}]+/))
```

> What you're trying to do is very standard and there are similar constructs in most of the existing grammars.

That encouraging to hear, I will take a look at YAML and Python, both of which I know has some similar concepts to what I'm trying to do.

**Reply by maxbrunsfeld** [reply]
Ok, that makes sense. Now you're getting to the hard part 😁 . This `unquoted_string` rule seems pretty fundamentally ambiguous, so it might take some more experimentation to find out what are the actual behaviors of the language that you're modeling.

You might have to have more special cases in your regex. Or sometimes, for this type of thing, which is difficult to express in terms of a regex, you may want to use an external scanner, so that you can apply arbitrary imperative logic for deciding what should be interpreted as an `unquoted_string`.

**Reply by maxbrunsfeld** [reply]
I might actually look at the `bash` grammar as a guide, because bash has a lot of strange lexical rules for literal words.

**Reply by Hubro** [reply]
@maxbrunsfeld I just had a brain blast of inspiration. If I give `$.unquoted_string` low lexical precedence, it means that Tree-sitter will lex any other defined token first, resulting in a syntax tree like:

```c
// These should all be parsed as (argument (unquoted_string))

default 127.0.0.1;   // (ERROR statement (keyword) (argument (integer) (GARBAGE))
default 0x1234_foo;    // (ERROR statement (keyword) (argument (hex) (GARBAGE))
default foo^bar;   // (ERROR statement (keyword) (argument (node_identifier) (GARBAGE))
default true&fair;   // (ERROR statement (keyword) (argument (boolean) (GARBAGE))
```

So the solution is just to define an abomination of a rule for `$.unquoted_string` that allows either an unquoted string, or any other argument type followed by an unquoted string:

```javascript
_unquoted_string: $ => token(prec(-1, /[^\s;"'{}]+/)),

unquoted_string: $ => choice(
    $._unquoted_string,
    seq(
        choice(
            $._node_identifier,
            $._integer,
            $._hex,
            $._boolean,
            $._keypath,
        ),
        $._unquoted_string,
    ),
),
```

And also I have to define a hidden version of each of my argument types:

```javascript
_node_identifier: $ => node_identifier,

node_identifier: $ => node_identifier,

_integer: $ => /-?\d+/,

integer: $ => $._integer,

_hex: $ => /-?0[xX][a-zA-Z0-9]+/,

hex: $ => $._hex,

_boolean: $ => choice("true", "false"),

boolean: $ => choice("true", "false"),

// ...
```

It definitely feels like a dirty hack around the lexer, but it works! All tests are green! :tada: 

![woooo](https://media.tenor.com/images/780e5fedbbb97b37b52c84f318f60f6a/tenor.gif)

**Reply by maxbrunsfeld** [reply]
I like it. Nice work.
